### PR TITLE
Fast tokenizers

### PR DESCRIPTION
**Pull request by n1t0**
I am opening this PR to track the integration of `tokenizers`. 

At the moment, we created two new classes to represent the fast version of both GPT2 and Bert tokenizers. There are a few breaking changes compared to the current `GPT2Tokenizer` and `BertTokenizer`:
  - `add_special_token` is now specified during initialization
  - truncation and padding options are also setup during initialization

By default, `encode_batch` pads everything using the longest sequence, and `encode` does not pad at all. If `pad_to_max_length=True`, then we pad everything using this length.
If a `max_length` is specified, then everything is truncated according to the provided options. This should work exactly like before.

In order to try these, you must `pip install tokenizers` in your virtual env.

### CODE REVIEW

**Comment by julien-c** [comment]
Can you remind why you moved those options to initialization vs. at `encode` time?

Is that a hard requirement of the native implem?

**Comment by n1t0** [comment]
Sure! The native implementation doesn't have `kwargs` so we need to define a static interface with pre-defined function arguments. This means that the configuration of the tokenizer is done by initializing its various parts and attaching them. There would be some unwanted overhead in doing this every time we `encode`.
I think we generally don't need to change the behavior of the tokenizer while using it, so this shouldn't be a problem. Plus, I think it makes the underlying functions clearer, and easier to use.

File: transformers/tokenization_gpt2.py [hunk 1]
```diff
@@ -237,3 +237,36 @@ def save_vocabulary(self, save_directory):
...
+    def __init__(self, vocab_file, merges_file, unk_token="<|endoftext|>", bos_token="<|endoftext|>",
+                 eos_token="<|endoftext|>", pad_to_max_length=False, add_prefix_space=True,
+                 max_length=None, stride=0, truncation_strategy='longest_first', **kwargs):
+
+        try:
+            from tokenizers import Tokenizer, models, pre_tokenizers, decoders
+
+            super(GPT2TokenizerFast, self).__init__(bos_token=bos_token, eos_token=eos_token, unk_token=unk_token, **kwargs)
+
+            self._tokenizer = Tokenizer(models.BPE.from_files(vocab_file, merges_file))
```
Comment (julien-c) [inline transformers/tokenization_gpt2.py]: Finding a way to get inline documentation in the `tokenizers` module is a must IMO (maybe through `.pyi` files?)

File: transformers/tokenization_gpt2.py [hunk 2]
```diff
@@ -237,3 +237,36 @@ def save_vocabulary(self, save_directory):
...
+    vocab_files_names = VOCAB_FILES_NAMES
+    pretrained_vocab_files_map = PRETRAINED_VOCAB_FILES_MAP
+    max_model_input_sizes = PRETRAINED_POSITIONAL_EMBEDDINGS_SIZES
+
+    def __init__(self, vocab_file, merges_file, unk_token="<|endoftext|>", bos_token="<|endoftext|>",
+                 eos_token="<|endoftext|>", pad_to_max_length=False, add_prefix_space=True,
+                 max_length=None, stride=0, truncation_strategy='longest_first', **kwargs):
+
+        try:
+            from tokenizers import Tokenizer, models, pre_tokenizers, decoders
```
Comment (julien-c) [inline transformers/tokenization_gpt2.py]: should we do an `import tokenizers as tk` like numpy/tf/torch to make it clear that we're using methods from the external dependency?

**Comment by LysandreJik** [comment]
Testing the fast tokenizers is not trivial. Doing tokenizer-specific tests is alright, but as of right now the breaking changes make it impossible to implement the fast tokenizers in the common tests pipeline.

In order to do so, and to obtain full coverage of the tokenizers, we would have to either:
- Split the current common tests in actual unit tests, rather than integration tests. As the arguments are passed in the `from_pretrained` (for the rust tokenizers) method rather than the `encode` (for the python tokenizers) method, having several chained calls to `encode` with varying arguments implies a big refactor of each test to test the rust tokenizers as well. Splitting those into unit tests would ease this task.
- Copy/Paste the common tests into a separate file (not ideal).

Currently there are some integration tests for the rust GPT-2 and rust BERT tokenizers, which may offer sufficient coverage for this PR. We would need the aforementioned refactor to have full coverage, which can be attended to in a future PR.

File: transformers/tokenization_gpt2.py (hunk 1 above)
Comment (thomwolf) [reply transformers/tokenization_gpt2.py]: Yes, that would be really nice. It can be more tricky than expected sometimes (pytorch is still struggling with that...). Do you want to have a look maybe? Otherwise, I can try to dive into it.

File: transformers/tokenization_gpt2.py (hunk 2 above)
Comment (thomwolf) [reply transformers/tokenization_gpt2.py]: I would favor this as well.

**Comment by thomwolf** [comment]
I really like it. Great work @n1t0 

I think now we should try to work on the other (python) tokenizers and see if we can find a middle ground behavior where they can both behave rather similarly, in particular for tests.

Also an open question: should we keep the "slow" python tokenizers that are easy to inspect? Could make sense, maybe renaming them to `BertTokenizerPython` for instance.

File: src/transformers/tokenization_bert.py [hunk 3]
```diff
@@ -525,3 +527,68 @@ def _is_punctuation(char):
     if cat.startswith("P"):
         return True
     return False
+
+
+class BertTokenizerFast(FastPreTrainedTokenizer):
```
Comment (thomwolf) [inline src/transformers/tokenization_bert.py]: Nitpicking but I would call them `FastBertTokenizer` instead of `BertTokenizerFast` so we match the base class `FastPreTrainedTokenizer`.

Also we could add a `fast=True` option to `AutoTokenizer` that would load a `Fast` version if available.

**Comment by n1t0** [comment]
Ok, I had to rewrite the whole history after the few restructuration PRs that have been merged. 

Since Python 2 has been dropped, I added `tokenizers` as a dependency (It is working for Python 3.5+).
We should now be ready to merge!

We should clearly keep the Python tokenizers and deprecate them slowly. In the end, I don't mind keeping them, but I'd like to avoid having to maintain both, especially if their API differ.

File: src/transformers/tokenization_utils.py [hunk 4]
```diff
@@ -1410,3 +1410,147 @@ def clean_up_tokenization(out_string):
             .replace(" 're", "'re")
         )
         return out_string
+
+
+class FastPreTrainedTokenizer(PreTrainedTokenizer):
```
Comment (thomwolf) [inline src/transformers/tokenization_utils.py]: We should also have token string and id attributes like `pad_token` and - pad_token_id` that we use often.

We could do complex stuff but maybe a simple way to handle that would be to have a single configuration for models and tokenizers which would store these attributes anyway, as we've started discussing recently with @julien-c

File: src/transformers/tokenization_utils.py [hunk 5]
```diff
@@ -1410,3 +1410,147 @@ def clean_up_tokenization(out_string):
...
+
+    @property
+    def decoder(self):
+        if self._decoder is None:
+            raise NotImplementedError
+        return self._decoder
+
+    @property
+    def vocab_size(self):
+        return self.tokenizer.get_vocab_size(False)
```
Comment (thomwolf) [inline src/transformers/tokenization_utils.py]: Can we have a self explainable keyword argument here like `self.tokenizer.get_vocab_size(with_added_tokens=False)` or it's not possible?

File: src/transformers/tokenization_utils.py (hunk 4 above)
Comment (thomwolf) [reply src/transformers/tokenization_utils.py]: Open to other options though @n1t0

File: src/transformers/tokenization_utils.py [hunk 6]
```diff
@@ -1410,3 +1410,147 @@ def clean_up_tokenization(out_string):
...
+        }
+        if return_token_type_ids:
+            encoding_dict["token_type_ids"] = encoding.type_ids
+        if return_attention_mask:
+            encoding_dict["attention_mask"] = encoding.attention_mask
+        if return_overflowing_tokens:
+            overflowing = encoding.overflowing
+            encoding_dict["overflowing_tokens"] = overflowing.ids if overflowing is not None else []
+        if return_special_tokens_mask:
+            encoding_dict["special_tokens_mask"] = encoding.special_tokens_mask
```
Comment (thomwolf) [inline src/transformers/tokenization_utils.py]: Does `encoding` contain tokens to char/byte mappings as well?

We could take the occasion of the `tokenizers` project to add this features as well.
This is an important feature. Current status is that we are waiting to see how `tokenizers` will handle it to add matching python implementations for the non-fast tokenizers.

File: src/transformers/tokenization_utils.py [hunk 7]
```diff
@@ -1410,3 +1410,147 @@ def clean_up_tokenization(out_string):
...
+            return_attention_mask=return_attention_mask,
+            return_overflowing_tokens=return_overflowing_tokens,
+            return_special_tokens_mask=return_special_tokens_mask,
+        )
+
+    def tokenize(self, text):
+        return self.tokenizer.encode(text).tokens
+
+    def _convert_token_to_id_with_added_voc(self, token):
+        return self.tokenizer.token_to_id(token)
```
Comment (thomwolf) [inline src/transformers/tokenization_utils.py]: What happens here if the token is not in the vocabulary?

File: src/transformers/tokenization_utils.py (hunk 4 above)
Comment (n1t0) [reply src/transformers/tokenization_utils.py]: I'm not entirely sure to understand what you mean.

The attributes are inherited from `PreTrainedTokenizer` so it should be almost the same. The only problem at the moment is when we update these after initialization, it doesn't call `_update_special_tokens()`

File: src/transformers/tokenization_utils.py (hunk 5 above)
Comment (n1t0) [reply src/transformers/tokenization_utils.py]: Yes, well I have a question about this. I am actually wondering, when is it interesting to get the `vocab_size` without the added tokens?

File: src/transformers/tokenization_utils.py (hunk 7 above)
Comment (n1t0) [reply src/transformers/tokenization_utils.py]: It just returns `None`

File: src/transformers/tokenization_utils.py (hunk 6 above)
Comment (n1t0) [reply src/transformers/tokenization_utils.py]: Not at the moment. I have all the structure in place to handle this, but it doesn't always work for now. The GPT2Tokenizer is handled out of the box since there's no magic done, but I can't say the same for BertTokenizer. I'll need to handle the Normalization and PreTokenization steps properly first before we can expose the offsets to the user.

File: transformers/tokenization_gpt2.py (hunk 1 above)
Comment (n1t0) [reply transformers/tokenization_gpt2.py]: I'll look into it!

File: src/transformers/tokenization_bert.py (hunk 3 above)
Comment (n1t0) [reply src/transformers/tokenization_bert.py]: I thought it would be better to have `Fast` as a suffix because it would allow text editors to propose the fast version when a user starts typing `BertTokenizer` for example. I don't mind changing to `FastBertTokenizer` if you prefer this version, but I'd rather change to `PreTrainedTokenizerFast` to keep the mentioned benefit. What do you think?

Comment (thomwolf) [reply src/transformers/tokenization_bert.py]: Ok!

File: src/transformers/tokenization_utils.py (hunk 4 above)
Comment (thomwolf) [reply src/transformers/tokenization_utils.py]: Oh indeed, my bad

File: src/transformers/tokenization_utils.py (hunk 5 above)
Comment (thomwolf) [reply src/transformers/tokenization_utils.py]: Not in many cases I think. It was mostly because we handled the added tokens a bit differently than native tokens in python and they were added in a second time so I'll kept backward compatibility with previous code using `vocab_size`.

We can probably get rid of `vocab_size` if it's impacting the Rust code.

File: src/transformers/tokenization_utils.py (hunk 7 above)
Comment (thomwolf) [reply src/transformers/tokenization_utils.py]: For tokenizers like `BertTokenizer` (and actually all tokenizers except the GPT2 ones) we would like to return the unknown token in this case `"[UNK]"`. I don't think this is handled here, is it?

**Comment by n1t0** [comment]
This is ready for the final review!

**Comment by thomwolf** [comment]
Ok, this is great, merging!
